qiskit/extensions/quantum_initializer/diagonal.py

```diff
@@ -48,16 +48,28 @@ def __init__(self, diag):
         num_action_qubits = math.log2(len(diag))
         if num_action_qubits < 1 or not num_action_qubits.is_integer():
             raise QiskitError("The number of diagonal entries is not a positive power of 2.")
-        for z in diag:
+        diag = diag.copy()
+        only_complex = True
+        for i, z in enumerate(diag):
+            if np.isscalar(z) and z in (1.0, -1.0):
+                diag[i] = complex(z)
+                continue
             try:
-                complex(z)
+                if z is not complex:
+                    complex(z)
+                    only_complex = False
```

Comment:
I'm not sure I understand the point of this loop, it sets `only_complex` to False if the number is not complex. But even if it could be cast to complex, it would be false. Why not keep a global `try except` statement like?
```python
try:
   diag = [complex(z) for z in diag]
except:
  # error couldn't convert
```